tests/unit/ad/test_grid_operators.py

```diff
@@ -475,3 +476,172 @@ def test_divergence(mdg: pp.MixedDimensionalGrid, dim: int):
+    mdg.add_subdomains(grids)
+    for sd, data in mdg.subdomains(return_data=True):
+        data[pp.STATE] = {}
+        data[pp.PRIMARY_VARIABLES] = {}
+        for var in variables:
+            data[pp.PRIMARY_VARIABLES].update({var: {"cells": 1}})
+            data[pp.STATE][var] = np.random.rand(sd.num_cells)
+
+    # Ad boilerplate
+    dof_manager = pp.DofManager(mdg)
+    eq_manager = pp.ad.EquationManager(mdg, dof_manager)
+
+    # Standard Ad variables
+    ad_vars = [eq_manager.variable(g, var) for g in grids for var in variables]
+    # Merge variables over all grids
+    merged_vars = []
+    for var in variables:
+        gv = [(g, var) for g in grids]
+        merged_vars.append(eq_manager.merge_variables(gv))
+
+    # First check of standard variables. If this fails, something is really wrong
+    for sd in grids:
+        data = mdg.subdomain_data(sd)
+        for var in ad_vars:
+            if sd == var._g:
+                expr = var.evaluate(dof_manager)
+                # Check that the size of the variable is correct
+                assert np.allclose(expr.val, data[pp.STATE][var._name])
+                # Check that the Jacobian matrix has the right number of columns
+                assert expr.jac.shape[1] == dof_manager.num_dofs()
+
+    # Next, check that merged variables are handled correctly.
+    for var in merged_vars:
+        expr = var.evaluate(dof_manager)
+        vals = []
+        for sub_var in var.sub_vars:
+            vals.append(mdg.subdomain_data(sub_var._g)[pp.STATE][sub_var._name])
+
+        assert np.allclose(expr.val, np.hstack([v for v in vals]))
+        assert expr.jac.shape[1] == dof_manager.num_dofs()
+
+    # Finally, check that the size of the Jacobian matrix is correct when combining
+    # variables (this will cover both variables and merged variable with the same name,
+    # and with different name).
+    for sd in grids:
+        for var in ad_vars:
+            nc = var.size()
+            cols = np.arange(nc)
+            data = np.ones(nc)
+            for mv in merged_vars:
+                nr = mv.size()
+
+                # The variable must be projected to the full set of grid for addition
+                # to be meaningful. This requires a bit of work.
+                sv_size = np.array([sv.size() for sv in mv.sub_vars])
+                mv_grids = [sv._g for sv in mv.sub_vars]
+                ind = mv_grids.index(var._g)
+                offset = np.hstack((0, np.cumsum(sv_size)))[ind]
+                rows = offset + np.arange(nc)
+                P = pp.ad.Matrix(sps.coo_matrix((data, (rows, cols)), shape=(nr, nc)))
+
+                eq = eq = mv + P * var
+                expr = eq.evaluate(dof_manager)
+                # Jacobian matrix size is set according to the dof manager,
+                assert expr.jac.shape[1] == dof_manager.num_dofs()
+
+
+def test_ad_discretization_class():
+    # Test of the mother class of all discretizations (pp.ad.Discretization)
```

Comment:
Sorry for nagging about this, but I don't like having to write ad in pp.ad.Discretization

Reply:
My problem is not so much the extra ad, that we have inconsistencies between ad and the other modules, and that we generally have no system in which classes and functions should be exposed on the top level. 

I'm in favor of taking a new look at this, but right now there are too many other moving pieces.

Reply:
Agreed